### ISSUE TITLE
Keyword Mappings of limited value to HTML content

### ISSUE THREAD

**Issue by GavinOsborn** [auth0/auth0-deploy-cli#10]
**Scenario**
 - You use the deploy-cli to deploy your custom logon page
 - Your project promotes changes through a series of environments before getting to productioin
 - Each environment has a different CDN for hosting css, images, scripts, etc

Ideally, in this scenario you'd want to be able to build your page to accept a Keyword Mapping to replace the CDN host address during deployment.

Something like this:

```html
<html>
    <head>
   <script src="@@CDN_HOST@@/snazzy-login.js"></script>
   <link href="@@CDN_HOST@@/styles/login.css" rel="stylesheet" type="text/css"/>
    </head>
    <body>
      <h1> Logon! </h1>
      <img src="@@CDN_HOST@@/images/company_logo.png" />
    </body>
</html>
```
With a config file like this:
```json
{
  "AUTH0_DOMAIN": "...",
  "AUTH0_CLIENT_ID": "...",
  "AUTH0_CLIENT_SECRET": "...",
  "AUTH0_KEYWORD_REPLACE_MAPPINGS": {
    "CDN_HOST": "https://prod.cdn.com"
  }
}
```

Unfortunately this yields extra `"` in the HTML when deployed eg:
> href=""https://prod.cdn.com"/styles/login.css"

I assume this is because the mapping was originally written for building JSON trees like
```
{
  "host": @@CDN_HOST@@
}
```

My local testing is consistent, I'm surprised this didn't come up in #3 when @nickcoury was trying something similar?
> I'm trying to replace a keyword in an HTML page and for the life of me can't get it to work as the documentation shows.

Am I imagining things?

**Comment by mostekcm** [comment]
@GavinOsborn I have created a PR for the underlying library that needs to be updated before I can fix this: https://github.com/auth0-extensions/auth0-source-control-extension-tools/pull/17

Once that PR is pushed and the npm library is updated, I'll release a new version of the deploy CLI.

**Comment by mostekcm** [comment]
@GavinOsborn if you want to contribute.  I'm trying to decide which way is the best to handle adding this support.

Option 1:
```
"AUTH0_KEYWORD_REPLACEMENT_MAPPINGS": {
   "auth0-simple-mappings": {
     "simple-key": "simple-value"
  },
  "auth0-json-mappings": {
    "json-key": [ { "key1": "value1" }, { "key2", "value2" } ]
  }
}
```

Option 2:
```
"AUTH0_KEYWORD_REPLACEMENT_MAPPINGS": {
  "json-key": [ { "key1": "value1" }, { "key2", "value2" } ]
},
"AUTH0_KEYWORD_REPLACEMENT_MAPPINGS_SIMPLE": {
   "simple-key": "simple-value"
}
```

**Comment by mostekcm** [comment]
Even with Option 1, the old way will still work with JSON replacements if they don't specify the sub-mappings type.

e.g.:
```
"AUTH0_KEYWORD_REPLACEMENT_MAPPINGS": {
  "json-key": [ { "key1": "value1" }, { "key2", "value2" } ]
}
```

**Comment by GavinOsborn** [comment]
@mostekcm Of the two I prefer *Option 1* but I feel it could be streamlined a bit
What do you think of
```js
{
  "AUTH0_KEYWORD_REPLACEMENT_MAPPINGS": {
    "key-values": { // perhaps called string-values or literal-values?
      "simple-key": "simple-value"
    },
    "json-values": {
      "key1": "value1",
      "key2": "value2",
      "key3": ["value 3.1", "value 3.2"]
    }
  }
}
```

 - I feel the `auth0-` sub-property prefixes are unnecessary
 - I don't really see why the `json-keys/json-values` property was an array of objects.
Do we lose anything by boiling it down to an object?

**Comment by GavinOsborn** [comment]
Infact I prefer `literal-values` the two property names clearly describes how those values will get serialized into the template.

**Comment by GavinOsborn** [comment]
You will have to deal with conflicts in either-case.
What if you have the same key in each property bag?
 - Do you denote it in the template?
`@@KEY-VALUE@@` vs `@JSON-VALUE@`
 - Or do you handle it the tool and reject / define precedence?

**Comment by GavinOsborn** [comment]
(I would prefer taking control of the serialization in the template)

Retaining @@JSON-VALUE@@ for backwards compatability with current JSON serialization norm and introduce a new literal token:
`@LITERAL-VALUE@` OR `##LITERAL-VALUE##`?

**Comment by mostekcm** [comment]
@GavinOsborn love that last idea!  See this PR: https://github.com/auth0-extensions/auth0-source-control-extension-tools/pull/17

I've changed it to just support @@key@@ => JSON replace, ##key## => Literal replace

**Comment by mostekcm** [comment]
Keeps it much simpler!

**Comment by GavinOsborn** [comment]
Hey @mostekcm, this seems to have stalled.
Have we encountered a problem?

**Comment by mostekcm** [comment]
@GavinOsborn you are right, it is in holding waiting for me to get write permissions to NPM :)

You can clone the repo locally if you want to get the new functionality while we wait for my permissions to come through :).

**Comment by GavinOsborn** [comment]
Nothing desperate yet, just holding off CDN support in our project until this drops.

Assuming this isn't *that* far away we can still wait :-)

**Comment by mostekcm** [comment]
This is fixed as of version 1.0.10 just pushed to npm

**Comment by GavinOsborn** [comment]
Just shipped our adoption of this feature. Works a treat! Thanks!
